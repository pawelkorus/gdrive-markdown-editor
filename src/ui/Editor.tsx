import React from 'react'
import { useState } from 'react'
import { MarkdownContent } from "../markdown"
import { showPicker } from '../google'

type Props = {
    content: string,
    onSaveClicked?: (event:SaveEvent) => void
    onCloseClicked?: () => void
    onTogglePreviewClicked?: () => void
}

export class SaveEvent {
    constructor(
        public content:string
    ) {}
}

export default function(props:Props):React.ReactElement {
    const [previewEnabled, setPreviewEnabled] = useState(false)
    const [updatedContent, setUpdatedContent] = useState(props.content)

    function togglePreview() {
        setPreviewEnabled(!previewEnabled);
        props.onTogglePreviewClicked && props.onTogglePreviewClicked();
    }

    function browseGdrive() {
        showPicker().then(res => setUpdatedContent(updatedContent + res))
    }

    return (
<div>
    <div className="container-fluid p-2">
        <div className="d-flex flex-row">
            <div className="me-auto me-2"></div>
            <button className="btn btn-primary ms-1" id="btn-save" type="button" onClick={e => props.onSaveClicked(new SaveEvent(updatedContent))}>Save</button>
            <button className="btn btn-primary ms-1" id="btn-browse" type="button" onClick={browseGdrive}>Browse</button>
            <button className="btn btn-primary ms-1" id="btn-preview" type="button" onClick={togglePreview}>Preview</button>
            <button className="btn btn-primary ms-1" id="btn-close" type="button" onClick={props.onCloseClicked}>Close</button>
        </div>
    </div>
    <div className="container-lg mt-4" id="main">
        <div className="row content"> 
            <div className="col">
                    <textarea className="form-control w-100 h-100" 
                        value={updatedContent} 
                        onChange={e => { setUpdatedContent(e.target.value) }}></textarea>
            </div>

            {previewEnabled && <div className="col"><MarkdownContent content={updatedContent}/></div>}    
        </div>
    </div>
</div>
)}
