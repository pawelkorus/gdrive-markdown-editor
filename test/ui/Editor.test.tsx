import React from 'react';
import renderer, { act } from 'react-test-renderer';
import Editor, { Props } from '../../src/ui/Editor';
import { test, expect, jest, afterEach } from '@jest/globals';


// create jest snapshot serializer that will replace all date strings values with a placeholder. Date strings will have format of "Last saved at 1/1/2021, 12:00:00 AM"
const customSnapshotSerializer = {
    test: (val:any) => {
        if(val.type == 'small' && val.props.className == 'text-success') return true
    },
    print: (val:any) => '<small>Last saved at</small>'
};

expect.addSnapshotSerializer(customSnapshotSerializer);

let component: renderer.ReactTestRenderer;

afterEach(() => {
    component.unmount()
});

test('renders without crashing', () => {
    const props:Props = {
        content: '',
        onSaveClicked: () => {},
        onCloseClicked: () => {},
        onTogglePreviewClicked: () => {}
    };
    component = renderer.create(<Editor {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('renders with content', () => {
    const props:Props = {
        content: 'Hello, world!',
        onSaveClicked: () => {},
        onCloseClicked: () => {},
        onTogglePreviewClicked: () => {}
    };
    component = renderer.create(<Editor {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('renders with preview enabled', () => {
    const mockOnTogglePreviewClicked = jest.fn();
    const props:Props = {
        content: 'Hello, world!',
        onSaveClicked: () => {},
        onCloseClicked: () => {},
        onTogglePreviewClicked: mockOnTogglePreviewClicked
    };
    component = renderer.create(<Editor {...props} />);
    act(() => {
        component.root.findByProps({id: 'btn-preview'}).props.onClick();
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('save button click', () => {
    const props:Props = {
        content: 'Hello, world!',
        onSaveClicked: () => {},
        onCloseClicked: () => {},
        onTogglePreviewClicked: () => {}
    };
    component = renderer.create(<Editor {...props} />);
    act(() => {
        component.root.findByProps({id: 'btn-save'}).props.onClick();
    });

    const tree = component.toJSON();

    // tree.replace(/Last saved at .+/, 'Last saved at 1/1/2021, 12:00:00 AM')
    expect(tree).toMatchSnapshot();
});

test('close button click', () => {
    const mockOnCloseClicked = jest.fn();
    const props:Props = {
        content: 'Hello, world!',
        onSaveClicked: () => {},
        onCloseClicked: mockOnCloseClicked,
        onTogglePreviewClicked: () => {}
    };
    component = renderer.create(<Editor {...props} />);
    act(() => {
        component.root.findByProps({id: 'btn-close'}).props.onClick();
    });

    expect(mockOnCloseClicked).toHaveBeenCalled();
});

test('preview button click', () => {
    const mockOnTogglePreviewClicked = jest.fn();
    const props:Props = {
        content: 'Hello, world!',
        onSaveClicked: () => {},
        onCloseClicked: () => {},
        onTogglePreviewClicked: mockOnTogglePreviewClicked
    };
    component = renderer.create(<Editor {...props} />);
    act(() => {
        component.root.findByProps({id: 'btn-preview'}).props.onClick();
    });

    expect(mockOnTogglePreviewClicked).toHaveBeenCalled();
});


