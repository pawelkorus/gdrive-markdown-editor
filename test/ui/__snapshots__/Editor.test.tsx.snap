// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with content 1`] = `
<div
  className="container-fluid p-2 d-flex flex-column min-vh-100"
>
  <div
    className="d-flex flex-row align-items-center justify-content-end"
  >
    <h5
      className="me-auto mb-0"
      onClick={[Function]}
    >
      test.md
    </h5>
    <button
      className="btn btn-primary ms-1"
      id="btn-save"
      onClick={[Function]}
      type="button"
    >
      Save
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-browse"
      onClick={[Function]}
      type="button"
    >
      Browse
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-preview"
      onClick={[Function]}
      type="button"
    >
      Preview
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-close"
      onClick={[Function]}
      type="button"
    >
      Close
    </button>
  </div>
  <div
    className="d-flex flex-row flex-fill mt-4"
  >
    <div
      className="position-relative"
      style={
        {
          "flex": 1,
        }
      }
    >
      <textarea
        className="form-control position-absolute"
        onChange={[Function]}
        style={
          {
            "minHeight": "100%",
          }
        }
        value="Hello, world!"
      />
    </div>
  </div>
</div>
`;

exports[`renders with preview enabled 1`] = `
<div
  className="container-fluid p-2 d-flex flex-column min-vh-100"
>
  <div
    className="d-flex flex-row align-items-center justify-content-end"
  >
    <h5
      className="me-auto mb-0"
      onClick={[Function]}
    >
      test.md
    </h5>
    <button
      className="btn btn-primary ms-1"
      id="btn-save"
      onClick={[Function]}
      type="button"
    >
      Save
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-browse"
      onClick={[Function]}
      type="button"
    >
      Browse
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-preview"
      onClick={[Function]}
      type="button"
    >
      Preview
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-close"
      onClick={[Function]}
      type="button"
    >
      Close
    </button>
  </div>
  <div
    className="d-flex flex-row flex-fill mt-4"
  >
    <div
      className="position-relative"
      style={
        {
          "flex": 1,
        }
      }
    >
      <textarea
        className="form-control position-absolute"
        onChange={[Function]}
        style={
          {
            "minHeight": "100%",
          }
        }
        value="Hello, world!"
      />
    </div>
    <div
      className="p-2"
      style={
        {
          "flex": 1,
        }
      }
    >
      <p>
        Hello, world!
      </p>
    </div>
  </div>
</div>
`;

exports[`renders without crashing 1`] = `
<div
  className="container-fluid p-2 d-flex flex-column min-vh-100"
>
  <div
    className="d-flex flex-row align-items-center justify-content-end"
  >
    <h5
      className="me-auto mb-0"
      onClick={[Function]}
    >
      test.md
    </h5>
    <button
      className="btn btn-primary ms-1"
      id="btn-save"
      onClick={[Function]}
      type="button"
    >
      Save
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-browse"
      onClick={[Function]}
      type="button"
    >
      Browse
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-preview"
      onClick={[Function]}
      type="button"
    >
      Preview
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-close"
      onClick={[Function]}
      type="button"
    >
      Close
    </button>
  </div>
  <div
    className="d-flex flex-row flex-fill mt-4"
  >
    <div
      className="position-relative"
      style={
        {
          "flex": 1,
        }
      }
    >
      <textarea
        className="form-control position-absolute"
        onChange={[Function]}
        style={
          {
            "minHeight": "100%",
          }
        }
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`save button click 1`] = `
<div
  className="container-fluid p-2 d-flex flex-column min-vh-100"
>
  <div
    className="d-flex flex-row align-items-center justify-content-end"
  >
    <h5
      className="me-auto mb-0"
      onClick={[Function]}
    >
      test.md
    </h5>
    <span
      className="ms-1"
    >
      <small>Last saved at</small>
    </span>
    <button
      className="btn btn-primary ms-1"
      id="btn-save"
      onClick={[Function]}
      type="button"
    >
      Save
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-browse"
      onClick={[Function]}
      type="button"
    >
      Browse
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-preview"
      onClick={[Function]}
      type="button"
    >
      Preview
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-close"
      onClick={[Function]}
      type="button"
    >
      Close
    </button>
  </div>
  <div
    className="d-flex flex-row flex-fill mt-4"
  >
    <div
      className="position-relative"
      style={
        {
          "flex": 1,
        }
      }
    >
      <textarea
        className="form-control position-absolute"
        onChange={[Function]}
        style={
          {
            "minHeight": "100%",
          }
        }
        value="Hello, world!"
      />
    </div>
  </div>
</div>
`;

exports[`should enter file name edit mode 1`] = `
<div
  className="container-fluid p-2 d-flex flex-column min-vh-100"
>
  <div
    className="d-flex flex-row align-items-center justify-content-end"
  >
    <input
      autoFocus={true}
      className="form-control me-auto"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="test.md"
    />
    <button
      className="btn btn-primary ms-1"
      id="btn-save"
      onClick={[Function]}
      type="button"
    >
      Save
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-browse"
      onClick={[Function]}
      type="button"
    >
      Browse
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-preview"
      onClick={[Function]}
      type="button"
    >
      Preview
    </button>
    <button
      className="btn btn-primary ms-1"
      id="btn-close"
      onClick={[Function]}
      type="button"
    >
      Close
    </button>
  </div>
  <div
    className="d-flex flex-row flex-fill mt-4"
  >
    <div
      className="position-relative"
      style={
        {
          "flex": 1,
        }
      }
    >
      <textarea
        className="form-control position-absolute"
        onChange={[Function]}
        style={
          {
            "minHeight": "100%",
          }
        }
        value="Hello, world!"
      />
    </div>
  </div>
</div>
`;
